brozzler/ydl.py

```diff
@@ -295,58 +326,81 @@ def _remember_videos(page, pushed_videos=None):
-            )
-        return ie_result
-    except brozzler.ShutdownRequested as e:
-        raise
-    except Exception as e:
-        if hasattr(e, "exc_info") and e.exc_info[0] == yt_dlp.utils.UnsupportedError:
-            return None
-        elif (
-            hasattr(e, "exc_info")
-            and e.exc_info[0] == urllib.error.HTTPError
-            and hasattr(e.exc_info[1], "code")
-            and e.exc_info[1].code == 420
-        ):
-            raise brozzler.ReachedLimit(e.exc_info[1])
-        elif (
-            hasattr(e, "exc_info")
-            and e.exc_info[0] == urllib.error.URLError
-            and worker._proxy_for(site)
-        ):
-            # connection problem when using a proxy == proxy error (XXX?)
-            raise brozzler.ProxyError(
-                "yt-dlp hit apparent proxy error from " "%s" % ytdlp_url
-            ) from e
-        else:
+    attempt = 0
+    while attempt < MAX_YTDLP_ATTEMPTS:
+        try:
+            logging.info("trying yt-dlp on %s", ydl.url)
+            # should_download_vid = not ydl.youtube_host
+            # then
+            # ydl.extract_info(str(urlcanon.whatwg(ydl.url)), download=should_download_vid)
+            # if ydl.youtube_host and ie_result:
+            #     download_url = ie_result.get("url")
+            metrics.brozzler_ydl_extract_attempts.labels(ydl.youtube_host).inc(1)
+            with brozzler.thread_accept_exceptions():
+                # we do whatwg canonicalization here to avoid "<urlopen error
+                # no host given>" resulting in ProxyError
+                # needs automated test
+                # and yt-dlp needs sanitize_info for extract_info
+                ie_result = ydl.sanitize_info(
+                    ydl.extract_info(str(urlcanon.whatwg(ydl.url)))
+                )
+            metrics.brozzler_ydl_extract_successes.labels(ydl.youtube_host).inc(1)
+            break
+        except brozzler.ShutdownRequested as e:
             raise
+        except Exception as e:
+            if (
+                hasattr(e, "exc_info")
+                and e.exc_info[0] == yt_dlp.utils.UnsupportedError
+            ):
+                return None
+            elif (
+                hasattr(e, "exc_info")
+                and e.exc_info[0] == urllib.error.HTTPError
+                and hasattr(e.exc_info[1], "code")
+                and e.exc_info[1].code == 420
+            ):
+                raise brozzler.ReachedLimit(e.exc_info[1])
+            else:
+                # todo: other errors to handle separately?
+                # OSError('Tunnel connection failed: 464 Host Not Allowed') (caused by ProxyError...)
+                # and others...
+                attempt += 1
+                if attempt == MAX_YTDLP_ATTEMPTS:
+                    logging.warning(
+                        "Failed after %s attempts. Error: %s", MAX_YTDLP_ATTEMPTS, e
+                    )
+                    raise brozzler.ProxyError(
```

Comment:
retryable yt-dlp proxy error... maybe subclass brozzler.ProxyError? 

also, need to avoid tight re-try loop